### PR TITLE
chore: add channel between file aggregation and load steps

### PR DESCRIPTION
**Pull request by tryangul**
## What
Reworks interface between file aggregation (`SpillToDiskTask`) and load stages (`ProcessRecordsTask`) to provide a clearer mechanism for limiting record loading concurrency and propagating backpressure 

## Details
* Adds `MultiProducerChannel` for buffering ("fan in") between `SpillToDiskTask`s (1 per stream) and the limited `ProcessRecordsTask`s—named `fileAggregateQueue`
* Adds config to limit the number of concurrent `ProcessRecordsTask`
* `SpillToDiskTask` now publishes to the shared `fileAggregateQueue` instead of calling back to the `DestinationTaskLauncher`
* Organizes `SpillToDiskTask` for easier reading / documentation

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/MultiProducerChannel.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import java.lang.IllegalStateException
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicLong
+import kotlinx.coroutines.channels.Channel
+
+/**
+ * A channel designed for use with a dynamic amount of producers. Close will only close the
+ * underlying channel, when there are no remaining registered producers.
+ */
+class MultiProducerChannel<T>(override val channel: Channel<T>) : ChannelMessageQueue<T>() {
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/MultiProducerChannel.kt:17]: If this pattern repeats, we could upgrade this into something that wraps any closable.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/implementor/ProcessRecordsTask.kt [hunk 2]
```diff
@@ -37,81 +38,84 @@ interface ProcessRecordsTask : StreamLevel, ImplementorScope
...
 
 interface ProcessRecordsTaskFactory {
     fun make(
         taskLauncher: DestinationTaskLauncher,
-        stream: DestinationStream.Descriptor,
-        file: SpilledRawMessagesLocalFile,
     ): ProcessRecordsTask
 }
 
+data class FileAggregateMessage(
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/implementor/ProcessRecordsTask.kt:97]: Looking ahead to possible interface changes, if `FileAggregateMessage` extends something like

```
interface Aggregate {
  records: Flow<Record>
}
```

Then the `SpillToDiskTask` can become something like `RecordAggregatorTask`, which gets injected with `RecordAggregator`, which could  be `FileAggregator` (the current behavior) or just a Noop. (Promoting an `Iterator` to a `Flow` is trivial.)

The trigger for that change could be "we want increased performance on the tasks that don't need buffering, like object storage uploads with small part sizes."

Then `ProcessRecordsTask` could become something like `AccumulateBatchTask`, which takes a `BatchAccumulator<T, U: Batch>` where we set up `nProcessRecordsTasks` instances of `BatchAccumulator<Aggregate, [client-defined Batch type]>` and `nProcessBatchTasks` of `BatchAccumulator<[client-defined], [client-defined]>`. (All of which just use the client-provided accumulators.)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt [hunk 3]
```diff
@@ -20,114 +25,170 @@ import io.airbyte.cdk.load.state.TimeWindowTrigger
...
-                                }
-                            }
-                        }
+        val initialAccumulator = createNewAccumulator()
+
+        val registration = outputQueue.registerProducer()
+        registration.use {
+            inputQueue.consume().fold(initialAccumulator) { acc, reserved ->
+                reserved.use {
+                    when (val event = it.value) {
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt:70]: I appreciate the clean-up here.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt [hunk 4]
```diff
@@ -20,114 +25,170 @@ import io.airbyte.cdk.load.state.TimeWindowTrigger
...
+                SpilledRawMessagesLocalFile(
+                    spillFile,
+                    sizeBytes,
+                    nextRange,
+                    endOfStream = true,
+                )
+
+            publishFile(file)
+        }
+        return FileAccumulator(
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt:148]: nit: we could use a comment here to flag that this is unused right now.

File: airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/message/MultiProducerChannelTest.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    fun `closes underlying channel when no producers are registered`() = runTest {
+        channel.registerProducer()
+        channel.registerProducer()
+        channel.registerProducer()
+
+        channel.close()
+        channel.close()
+        channel.close()
+        coVerify(exactly = 1) { wrapped.close() }
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/message/MultiProducerChannelTest.kt:54]: The `exactly = #` is a nice feature.

**Review by johnny-schmidt — APPROVED** [review]
LGTM, pending the time window fixes.

**Review by natikgadzhi — APPROVED** [review]
YAWLAW

**Pull request merged**